crates/ra_ide/src/hover.rs

```diff
@@ -335,6 +346,277 @@ fn hover_for_definition(db: &RootDatabase, def: Definition) -> Option<Markup> {
     }
 }
 
+// Rewrites a markdown document, resolving links using `callback` and additionally striping prefixes/suffixes on link titles.
+fn map_links<'e>(
+    events: impl Iterator<Item = Event<'e>>,
+    callback: impl Fn(&str, &str) -> (String, String),
+) -> impl Iterator<Item = Event<'e>> {
+    let mut in_link = false;
+    let mut link_target: Option<CowStr> = None;
+
+    events.map(move |evt| match evt {
+        Event::Start(Tag::Link(_link_type, ref target, _)) => {
+            in_link = true;
+            link_target = Some(target.clone());
+            evt
+        }
+        Event::End(Tag::Link(link_type, _target, _)) => {
+            in_link = false;
+            Event::End(Tag::Link(link_type, link_target.take().unwrap(), CowStr::Borrowed("")))
+        }
+        Event::Text(s) if in_link => {
+            let (link_target_s, link_name) = callback(&link_target.take().unwrap(), &s);
+            link_target = Some(CowStr::Boxed(link_target_s.into()));
+            Event::Text(CowStr::Boxed(link_name.into()))
+        }
+        Event::Code(s) if in_link => {
+            let (link_target_s, link_name) = callback(&link_target.take().unwrap(), &s);
+            link_target = Some(CowStr::Boxed(link_target_s.into()));
+            Event::Code(CowStr::Boxed(link_name.into()))
+        }
+        _ => evt,
+    })
+}
+
+/// Rewrite documentation links in markdown to point to an online host (e.g. docs.rs)
+fn rewrite_links(db: &RootDatabase, markdown: &str, definition: &Definition) -> String {
+    let doc = Parser::new_with_broken_link_callback(
+        markdown,
+        Options::empty(),
+        Some(&|label, _| Some((/*url*/ label.to_string(), /*title*/ label.to_string()))),
+    );
+
+    let doc = map_links(doc, |target, title: &str| {
+        // This check is imperfect, there's some overlap between valid intra-doc links
+        // and valid URLs so we choose to be too eager to try to resolve what might be
+        // a URL.
+        if target.contains("://") {
+            (target.to_string(), title.to_string())
+        } else {
+            // Two posibilities:
+            // * path-based links: `../../module/struct.MyStruct.html`
+            // * module-based links (AKA intra-doc links): `super::super::module::MyStruct`
+            let resolved = try_resolve_intra(db, definition, title, &target).or_else(|| {
+                try_resolve_path(db, definition, &target).map(|target| (target, title.to_string()))
```

Comment:
```suggestion
                try_resolve_legacy_link(db, definition, &target).map(|target| (target, title.to_string()))
```

Reply:
I'm not sure I like using legacy here, to me it almost implies deprecated, which these aren't. Happy to change it if you think it's better though.

Reply:
Hm, right, no one is going to deprecate URL :thinking: 

Maybe `try_guess_from_url`?

Reply:
I still think this name is the best choice. At this point the input is a filesystem path which just happens to get interpreted as part of a url.

This method is internal anyway and doesn't make up a public interface.